File: random_rotation.py
import numpy as np
from matplotlib import pyplot
from mpl_toolkits.mplot3d import Axes3D


def visualize(points):
    """
    points: n x 3
    """
    xs = points[:, 0]
    ys = points[:, 1]
    zs = points[:, 2]

    fig = pyplot.figure()
    ax = Axes3D(fig)

    ax.scatter(xs, ys, zs)
    pyplot.show()


if __name__ == '__main__':

    points = []
    for _ in range(1000):
        # https://math.stackexchange.com/a/1602779/209055
        q, r = np.linalg.qr(np.random.randn(3, 3))
        q[:, 0] = q[:, 0] * (1.0 if np.random.rand() < 0.5 else -1.0)
        q[:, 1] = np.linalg.det(q) * q[:, 1]
        point = q @ np.array([1, 0, 0])
        points.append(point)
    points = np.array(points)

    visualize(points)

File: random_rotation.m
for i=1:1000;
  [Q,~] = qr(randn(3));
  Q(:,1)=Q(:,1)*(2*(rand>0.5)-1);
  Q(:,2)=det(Q)*Q(:,2);
  dirs(:,i) = Q*[1 0 0]';
end;

scatter3(dirs(1,:), dirs(2,:), dirs(3,:));
axis equal vis3d;
